### ISSUE TITLE
Slow creation of CAs

### ISSUE THREAD

**Issue by dcabecinhas** [jonniedie/ComponentArrays.jl#48]
It seems the slow (as per v0.50 release notes) kwargs constructor is being used for all CA creation and not just for the CA(...; x=y) syntax.

I'm not sure if this was the result of a recent change or if it has always been like this.

```julia
julia> @btime ComponentArray(a=5, b=[1, 2])
  23.316 μs (45 allocations: 2.14 KiB)
ComponentVector{Int64}(a = 5, b = [1, 2])
```

```julia
julia> @code_warntype ComponentArray(a=5, b=[1, 2])
Variables
  #unused#::Core.Compiler.Const(Core.var"#Type##kw"(), false)
  @_2::NamedTuple{(:a, :b),Tuple{Int64,Array{Int64,1}}}
  @_3::Type{ComponentArray}
  kwargs...::Base.Iterators.Pairs{Symbol,Any,Tuple{Symbol,Symbol},NamedTuple{(:a, :b),Tuple{Int64,Array{Int64,1}}}}

Body::Any
1 ─      (kwargs... = Base.pairs(@_2))
│   %2 = ComponentArrays.:(var"#ComponentArray#13")(kwargs..., @_3)::Any
└──      return %2
```

**Comment by jonniedie** [comment]
Oh yeah, it’s for all keyword constructors. Not sure why I specifically pointed it out there.

As far as fixing the issue, the few things I’ve tried haven’t helped, but I have some other ideas I want to try soon.

**Comment by roflmaostc** [comment]
That's still an issue, isn't it?

I might want to use the creation of the CAs during the forward pass of the function and hence it needs to be rather efficient.


```julia
julia> @code_warntype ComponentArray((;a=1,b=[1,2,3,3]))
MethodInstance for ComponentArray(::NamedTuple{(:a, :b), Tuple{Int64, Vector{Int64}}})
  from ComponentArray(nt::NamedTuple) in ComponentArrays at /home/fxw/.julia/packages/ComponentArrays/mdPcX/src/componentarray.jl:63
Arguments
  #self#::Type{ComponentArray}
  nt::NamedTuple{(:a, :b), Tuple{Int64, Vector{Int64}}}
Body::Any
1 ─ %1 = ComponentArrays.make_carray_args(nt)::Tuple{Any, AbstractAxis}
│   %2 = Core._apply_iterate(Base.iterate, ComponentArrays.ComponentArray, %1)::Any
└──      return %2


julia> @time ComponentArray((;a=1,b=[1,2,3,3]))
  0.000050 seconds (47 allocations: 2.312 KiB)
ComponentVector{Int64}(a = 1, b = [1, 2, 3, 3])
```

**Comment by jonniedie** [comment]
There may be some room for improvement, but this is fundamentally type-unstable operation so it will never be truly fast using this style of constructor. To see this, consider the above example:
```julia
julia> ca = ComponentArray(a=1,b=[1,2,3,3])
ComponentVector{Int64}(a = 1, b = [1, 2, 3, 3])

julia> typeof(ca)
ComponentVector{Int64, Vector{Int64}, Tuple{Axis{(a = 1, b = 2:5)}}}
```
The `Axis{(a = 1, b = 2:5)}` part of this depends on the length of the `b` component, which isn't known at compile time. I think there's not really any way around this without making the operation of `ComponentArray`s slower. But it's possible that there are at least some ways around this to make it a bit faster, even if it's not type unstable. I'm open to suggestions.

**Comment by ToucheSir** [comment]
For my own curiosity, what benefits are gained by having `Axis{(a = 1, b = 2:5)}}` instead of `Axis{NamedTuple{(:a, :b), Tuple{Int64, UnitRange{Int64}}}}`?

**Comment by jonniedie** [comment]
`Axis{(a = 1, b = 2:5)}}` encodes the information about the indexing in the type domain. The other one, which we would have if `Axis` was a type that held the indexing `NamedTuple` as a field, would make the property access happen at runtime. Every version of the second one I tried was really slow due to type instability. At some point I should pick it up again and really give it a good effort because it would be pretty nice if we could get it working.

**Comment by ToucheSir** [comment]
Makes sense. I would've thought const prop could help with the second case, but maybe it balks on the recursion.

**Comment by jonniedie** [comment]
Yeah, I really would have thought so too. And who knows, maybe it does these days. I can’t remember the last time I tried it. Maybe around v1.5 or so.
